lib/integrations/azureoidc/accessgraph.go

```diff
@@ -0,0 +1,158 @@
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package azureoidc
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/json"
+	"io"
+	"log/slog"
+	"os"
+	"path"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/types"
+)
+
+// singleSignOnMode represents the possible values for `currentSingleSignOnMode` in `adSingleSignOn`
+type singleSignOnMode string
+
+const (
+	// singleSignOnModeNone indicates that the application does not have SSO set up.
+	singleSignOnModeNone singleSignOnMode = "none" //nolint:unused // this serves as documentation of a possible value.
+	// singleSignOnModeFederated indicates federated SSO such as SAML.
+	singleSignOnModeFederated singleSignOnMode = "federated"
+)
+
+// adSingleSignOn represents the response from https://main.iam.ad.ext.azure.com/api/ApplicationSso/{servicePrincipalID}/SingleSignOn
+type adSingleSignOn struct {
+	CurrentSingleSignOnMode singleSignOnMode `json:"currentSingleSignOnMode"`
+}
+
+// tagInfoCache is the format for the file produced by CreateTAGCacheFile.
+type tagInfoCache struct {
+	AppSsoSettingsCache []*types.PluginEntraIDAppSSOSettings `json:"app_sso_settings_cache"`
+}
+
+// getSingleSignOn uses Azure private API to get basic information about an enterprise applications single sign on mode.
+func getSingleSignOn(ctx context.Context, token string, servicePrincipalID string) (*adSingleSignOn, error) {
+	payload, err := privateAPIGet(ctx, token, path.Join("ApplicationSso", servicePrincipalID, "SingleSignOn"))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var result adSingleSignOn
+	if err := json.Unmarshal(payload, &result); err != nil {
+		return nil, trace.Wrap(err, "failed to deserialize SingleSignOn")
+	}
+
+	return &result, nil
+}
+
+// CreateTAGCacheFile populates a file containing the information necessary for Access Graph to analyze Azure SSO.
+func CreateTAGCacheFile(ctx context.Context) error {
+	graphClient, err := createGraphClient()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Get information about enterprise apps
+	appResp, err := graphClient.Applications().Get(ctx, nil)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Authorize to the private API
+	tenantID, err := getTenantID()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	token, err := getPrivateAPIToken(ctx, tenantID)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	cache := &tagInfoCache{}
+
+	for _, app := range appResp.GetValue() {
+		appID := app.GetAppId()
+		if appID == nil {
+			slog.WarnContext(ctx, "app ID is nil", "app", app)
+			continue
+		}
+		sp, err := graphClient.ServicePrincipalsWithAppId(appID).Get(ctx, nil)
+		if err != nil {
+			slog.ErrorContext(ctx, "could not retrieve service principal", "app_id", *appID, "error", err)
```

Comment:
This looks like a bug. If the error is not nil we should not try to call `GetId()` a few lines below.